tour_model_eval/get_scores.py

```diff
@@ -0,0 +1,44 @@
+import pandas as pd
+from pandas.testing import assert_frame_equal
+import label_processing as label_pro
+from sklearn import metrics
+import itertools
+
+
+# compare the trip orders in bin_trips with those in filter_trips above cutoff
+def compare_trip_orders(bins,bin_trips,filter_trips):
+    bin_trips_ts = pd.DataFrame(data=[trip["data"]["start_ts"] for trip in bin_trips])
+    bin_ls = list(itertools.chain(*bins))
+    bins_ts = pd.DataFrame(data=[filter_trips[i]["data"]["start_ts"] for i in bin_ls])
+    # compare two data frames, the program will continue to score calculation if two data frames are the same
+    assert_frame_equal(bins_ts, bin_trips_ts)
+
+
+# This function is to get homogeneity score after the first/second round of clustering
+# It is based on bin_trips, which are common trips. bin_trips are collected according to the indices in bins
+# More info about bin_trips is in similarity.py (delete_bins)
+def score(bin_trips, labels_pred):
+    bin_trips_user_input_df = pd.DataFrame(data=[trip["data"]["user_input"] for trip in bin_trips])
+    bin_trips_user_input_df = label_pro.map_labels(bin_trips_user_input_df)
+
+    # turn all user_input into list without binning
+    bin_trips_user_input_ls = bin_trips_user_input_df.values.tolist()
+    # drop duplicate user_input
+    no_dup_df = bin_trips_user_input_df.drop_duplicates()
+    # turn non-duplicate user_input into list
+    no_dup_list = no_dup_df.values.tolist()
+
+    # collect labels_true based on user_input
+    # To compute labels_true, we need to find out non-duplicate user labels, and use the index of the unique user label
+    # to label the whole trips
+    # if user labels are [home, work, home, home, work, exercise], the unique label list is [0,1,2],
+    # labels_true will be [0,1,0,0,1,2]
+    # labels_pred is the flattened list of labels of all common trips, e.g.[1,1,11,12,13,22,23]
```

Comment:
I still don't understand this. Don't the labels in `labels_true` and `labels_pred` have to be the same for the homogeneity score to work?

I'm actually really unsure about how this score works. This seems to indicate that the labels don't matter as long as there is a unique mapping between them.

```
>>> import sklearn.metrics as skm
>>> skm.homogeneity_score(["A", "B", "C", "B", "C"], [0,1,2,1,2])
1.0
```

But wait, no matter what I do, I get 1.0? I would like to see some examples of how the score calculation works, maybe in an auxiliary notebook that can eventually get converted to unit tests?

```
>>> skm.homogeneity_score(["A", "B", "C", "B", "C"], [0,1,2,1,3])
1.0
>>> skm.homogeneity_score(["A", "B", "C", "B", "C"], [0,1,2,1,4])
1.0
```

Reply:
It also seems like it would be much simpler if you had the same labels for `labels_pred` and `labels_true`

Reply:
and if the actual labels don't matter anyway, why do you need to convert 'home','work' etc to Yet Another set of indices?

Reply:
Same behavior if you change the first set of labels as well

```
>>> skm.homogeneity_score(["A", "B", "D", "B", "C"], [0,1,10,1,4])
1.0
>>> skm.homogeneity_score(["A", "B", "C", "B", "C"], [0,1,10,1,4])
1.0
```

Reply:
`labels_true` and `labels_pred` have to be numeric. It doesn't work when you pass in non-numeric values. I think that is all ML algorithms requires, right?
[Here is the documentation.](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.homogeneity_score.html)
I don't think we should explain everything in the py file that's well documented. 
It says - This metric is independent of the absolute values of the labels: a permutation of the class or cluster label values won’t change the score value in any way.
I have tried the labels and score for many times with examples when I learned how they used the labels to compute the score. There is no way to make `labels_pred` and `labels_true` with same labels. We have [home, work, exercise] as true user labels. But we may have [0,1,2] or [3,4,5] or other numeric labels after clustering. We don't need to make them exactly the same. `homogeneity_score` only sees the position of the labels.

Reply:
> It doesn't work when you pass in non-numeric values.

I don't find any evidence for this. It seems to work just fine with strings, as I showed above. Can you show that it doesn't work with strings?

Reply:
> There is no way to make `labels_pred` and `labels_true` with same labels.

I don't believe that it is impossible.
And if it were, we should abandon the project right now because eventually, we do need to make `labels_pred` map to the actual user labels.
It is not every useful to tell users that their trip was in cluster 5.
We have to tell them that the trip was for `home` or `work` or ....

Reply:
>  It says - This metric is independent of the absolute values of the labels: a permutation of the class or cluster label values won’t change the score value in any way.

I did consider the use case of a permutation - that is the first test that I did. What I don't understand is this

```
>>> skm.homogeneity_score(["A", "B", "C", "B", "C"], [0,1,2,1,4])
1.0
```

Label `C` maps to both `2` and `4`, but the score is still 1.0

Reply:
> Label C maps to both 2 and 4, but the score is still 1.0
If that is indeed the behavior of the homogeneity score, I am not sure it is a good metric for our use case.
Or at least, it needs stronger justification that it captures what we are looking for.

For this, the documentation has clearly explained.

> Non-perfect labelings that further split classes into more clusters can be perfectly homogeneous:
```
>>> print("%.6f" % homogeneity_score([0, 0, 1, 1], [0, 0, 1, 2]))
1.000000
>>> print("%.6f" % homogeneity_score([0, 0, 1, 1], [0, 1, 2, 3]))
1.000000
```
In your case, "C" is in index 2 and 4. In your `labels_pred`, index 2 and 4 can have labels 2 and 4 or 3 and 3. If they have same labels, that means they are in the same cluster, if not, it is still fine ( it means you have different clusters for same trips, e.g. 2 `home` trips with same modes in 2 clusters instead of in 1 cluster according to the Euclidean Distance). But if index 3, which is the same position as the second "B" in your `label_true`, also has label 2, the score will not be 1.0.
```
>>> skm.homogeneity_score(["A", "B", "C", "B", "C"], [0,1,2,2,4])
0.737175493807016
```

Reply:
@corinne-hcr I agree that the homogeneity score documentation has the explanation, but **you need to map it to this problem and justify the usage of this metric**. The explanation above about the home trips into two clusters is what I was looking for. I would also add the explanation of how we do capture that difference in our case (home trips split into two clusters will have the same homogeneity score but a lower request percentage)

Also, all your explanations above have focused on a single string - e.g. "home", but the user input is a tuple. You need to indicate how you are dealing with that.

Reply:
For mapping the labels, in one cluster, we do have trips with user labels. There are only two situations - trips are grouped in one cluster or the single trip in one cluster. Ideally, we should have labels [11,11,11] for trips in one cluster after the second round. In such a cluster, we should already have 1 or 2 trips that have user labels, the third one is predicted. Then we label it with the same user labels as the previous two. We don't have 1.0 homogeneity score for most of the cluster. It can also be [11,11,12]. Then for the third one, we will still label it with the same user labels as the previous two (since they are in one cluster).

Reply:
I don't see any of this explanation in the comments. Note that if people look at the code, they don't necessarily know which PR it came from and what the history in there is. You can add a link to this discussion if you like, but there *must* be at least a link to a high level description in a code comment.

Also, I don't see a follow up to my previous comments about labeling.
https://github.com/e-mission/e-mission-eval-private-data/pull/21#discussion_r642009487
https://github.com/e-mission/e-mission-eval-private-data/pull/21#discussion_r642009788